usr/src/man/man8/etrn.8
'\" te
.\" Copyright (c) 2004, Sun Microsystems, Inc.
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH ETRN 8 "November 22, 2021"
.SH NAME
etrn \- start mail queue run
.SH SYNOPSIS
.nf
\fBetrn\fR [\fB-b\fR] [\fB-v\fR] \fIserver-host\fR [\fIclient-hosts\fR]
.fi

.SH DESCRIPTION
SMTP's \fBETRN\fR command allows an SMTP client and server to interact, giving
the server an opportunity to start the processing of its queues for messages to
go to a given host. This is meant to be used in start-up conditions, as well as
for mail nodes that have transient connections to their service providers.
.sp
.LP
The \fBetrn\fR utility initiates an SMTP session with the host
\fIserver-host\fR and sends one or more \fBETRN\fR commands as follows: If no
\fIclient-hosts\fR are specified, \fBetrn\fR looks up every host name for which
\fBsendmail\fR(8) accepts email and, for each name, sends an \fBETRN\fR
command with that name as the argument. If any \fIclient-hosts\fR are
specified, \fBetrn\fR uses each of these as arguments for successive \fBETRN\fR
commands.
.SH OPTIONS
The following options are supported:
.sp
.ne 2
.na
\fB\fB-b\fR\fR
.ad
.RS 6n
System boot special case. Make sure localhost is accepting SMTP connections
before initiating the SMTP session with server-host.
.sp
This option is useful because it prevents race conditions between
\fBsendmail\fR(8) accepting connections and \fIserver-host\fR attempting to
deliver queued mail. This check is performed automatically if no
\fIclient-hosts\fR are specified.
.RE

.sp
.ne 2
.na
\fB\fB-v\fR\fR
.ad
.RS 6n
The normal mode of operation for \fBetrn\fR is to do all of its work silently.
The \fB-v\fR option makes it verbose, which causes \fBetrn\fR to display its
conversations with the remote SMTP server.
.RE

.SH ENVIRONMENT VARIABLES
No environment variables are used. However, at system start-up,
\fBsvc:/network/smtp:sendmail\fR reads \fB/etc/default/sendmail\fR. In this
file, if the variable \fBETRN_HOSTS\fR is set, \fBsvc:/network/smtp:sendmail\fR
parses this variable and invokes \fBetrn\fR appropriately. \fBETRN_HOSTS\fR
should be of the form:
.sp
.in +2
.nf
"s1:c1.1,c1.2        s2:c2.1 s3:c3.1,c3.2,c3.3"
.fi
.in -2
.sp

.sp
.LP
That is, white-space separated groups of \fIserver:client\fR where \fIclient\fR
can be one or more comma-separated names. The \fI:client\fR part is optional.
\fIserver\fR is the name of the server to prod; a mail queue run is requested
for each \fIclient\fR name. This is comparable to running:
.sp
.in +2
.nf
/usr/lib/sendmail -qR \fIclient\fR
.fi
.in -2
.sp

.sp
.LP
on the host \fIserver\fR.
.SH EXAMPLES
\fBExample 1 \fRUsing \fBetrn\fR
.sp
.LP
Inserting the line:

.sp
.in +2
.nf
ETRN_HOSTS="s1.example.com:clnt.example.com s2.example.com:clnt.example.com"
.fi
.in -2
.sp

.sp
.LP
in \fB/etc/default/sendmail\fR results in \fBsvc:/network/smtp:sendmail\fR
invoking \fBetrn\fR such that \fBETRN\fR commands are sent to both
\fBs1.example.com\fR and \fBs2.example.com\fR, with both having
\fBclnt.example.com\fR as the \fBETRN\fR argument.

.sp
.LP
The line:

.sp
.in +2
.nf
ETRN_HOSTS="server.example.com:client1.example.com,client2.example.com"
.fi
.in -2
.sp

.sp
.LP
results in two \fBETRN\fR commands being sent to \fBserver.example.com\fR, one
with the argument \fBclient1.example.com\fR, the other with the argument
\fBclient2.example.com\fR.

.sp
.LP
The line:

.sp
.in +2
.nf
ETRN_HOSTS="server1.example.com server2.example.com"
.fi
.in -2
.sp

.sp
.LP
results in set of a \fBETRN\fR commands being sent to both
\fBserver1.example.com\fR and \fBserver2.example.com\fR; each set contains one
\fBETRN\fR command for each host name for which \fBsendmail\fR(8) accepts
email, with that host name as the argument.

.SH FILES
.ne 2
.na
\fB\fB/etc/mail/sendmail.cf\fR\fR
.ad
.RS 25n
\fBsendmail\fR configuration file
.RE

.sp
.ne 2
.na
\fB\fB/etc/default/sendmail\fR\fR
.ad
.RS 25n
Variables used by \fBsvc:/network/smtp:sendmail\fR
.RE

.SH ATTRIBUTES
See \fBattributes\fR(7) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Interface Stability	Stable
.TE

.SH SEE ALSO
.BR attributes (7),
.BR sendmail (8)
.sp
.LP
RFC 1985
.SH NOTES
Not all SMTP servers support \fBETRN\fR.


usr/src/cmd/sendmail/aux/etrn.pl
#!/usr/perl5/bin/perl -w
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#
#
# Copyright (c) 1996-2000 by John T. Beck <john@beck.org>
# All rights reserved.
#
# Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
# Use is subject to license terms.
#

require 5.8.4;				# minimal Perl version required
use strict;
use warnings;
use English;

use Socket;
use Getopt::Std;
our ($opt_v, $opt_b);

# system requirements:
# 	must have 'hostname' program.

my $port = 'smtp';
select(STDERR);

chop(my $name = `hostname || uname -n`);

my ($hostname) = (gethostbyname($name))[0];

my $usage = "Usage: $PROGRAM_NAME [-bv] host [args]";
getopts('bv');
my $verbose = $opt_v;
my $boot_check = $opt_b;
my $server = shift(@ARGV);
my @hosts = @ARGV;
die $usage unless $server;
my @cwfiles = ();
my $alarm_action = "";

if (!@hosts) {
	push(@hosts, $hostname);

	open(CF, "</etc/mail/sendmail.cf") ||
	    die "open /etc/mail/sendmail.cf: $ERRNO";
	while (<CF>){
		# look for a line starting with "Fw"
		if (/^Fw.*$/) {
			my $cwfile = $ARG;
			chop($cwfile);
			my $optional = /^Fw-o/;
			# extract the file name
			$cwfile =~ s,^Fw[^/]*,,;

			# strip the options after the filename
			$cwfile =~ s/ [^ ]+$//;

			if (-r $cwfile) {
				push (@cwfiles, $cwfile);
			} else {
				die "$cwfile is not readable" unless $optional;
			}
		}
		# look for a line starting with "Cw"
		if (/^Cw(.*)$/) {
			my @cws = split (' ', $1);
			while (@cws) {
				my $thishost = shift(@cws);
				push(@hosts, $thishost)
				    unless $thishost =~ "$hostname|localhost";
			}
		}
	}
	close(CF);

	for my $cwfile (@cwfiles) {
		if (open(CW, "<$cwfile")) {
			while (<CW>) {
			        next if /^\#/;
				my $thishost = $ARG;
				chop($thishost);
				push(@hosts, $thishost)
				    unless $thishost =~ $hostname;
			}
			close(CW);
		} else {
			die "open $cwfile: $ERRNO";
		}
	}
	# Do this automatically if no client hosts are specified.
	$boot_check = "yes";
}

my ($proto) = (getprotobyname('tcp'))[2];
($port) = (getservbyname($port, 'tcp'))[2]
	unless $port =~ /^\d+/;

if ($boot_check) {
	# first connect to localhost to verify that we can accept connections
	print "verifying that localhost is accepting SMTP connections\n"
		if ($verbose);
	my $localhost_ok = 0;
	($name, my $laddr) = (gethostbyname('localhost'))[0, 4];
	(!defined($name)) && die "gethostbyname failed, unknown host $server";

	# get a connection
	my $sinl = sockaddr_in($port, $laddr);
	my $save_errno = 0;
	for (my $num_tries = 1; $num_tries < 5; $num_tries++) {
		socket(S, &PF_INET, &SOCK_STREAM, $proto)
			|| die "socket: $ERRNO";
		if (connect(S, $sinl)) {
			&alarm("sending 'quit' to $server");
			print S "quit\n";
			alarm(0);
			$localhost_ok = 1;
			close(S);
			alarm(0);
			last;
		}
		print STDERR "localhost connect failed ($num_tries)\n";
		$save_errno = $ERRNO;
		sleep(1 << $num_tries);
		close(S);
		alarm(0);
	}
	if (! $localhost_ok) {
		die "could not connect to localhost: $save_errno\n";
	}
}

# look it up

($name, my $thataddr) = (gethostbyname($server))[0, 4];
(!defined($name)) && die "gethostbyname failed, unknown host $server";

# get a connection
my $sinr = sockaddr_in($port, $thataddr);
socket(S, &PF_INET, &SOCK_STREAM, $proto)
	|| die "socket: $ERRNO";
print "server = $server\n" if (defined($verbose));
&alarm("connect to $server");
if (! connect(S, $sinr)) {
	die "cannot connect to $server: $ERRNO\n";
}
alarm(0);
select((select(S), $OUTPUT_AUTOFLUSH = 1)[0]);	# don't buffer output to S

# read the greeting
&alarm("greeting with $server");
while (<S>) {
	alarm(0);
	print if $verbose;
	if (/^(\d+)([- ])/) {
		# SMTP's initial greeting response code is 220.
		if ($1 != 220) {
			&alarm("giving up after bad response from $server");
			&read_response($2, $verbose);
			alarm(0);
			print STDERR "$server: NOT 220 greeting: $ARG"
				if ($verbose);
		}
		last if ($2 eq " ");
	} else {
		print STDERR "$server: NOT 220 greeting: $ARG"
			if ($verbose);
		close(S);
	}
	&alarm("greeting with $server");
}
alarm(0);
	
&alarm("sending ehlo to $server");
&ps("ehlo $hostname");
my $etrn_support = 0;
while (<S>) {
	if (/^250([- ])ETRN(.+)$/) {
		$etrn_support = 1;
	}
	print if $verbose;
	last if /^\d+ /;
}
alarm(0);

if ($etrn_support) {
	print "ETRN supported\n" if ($verbose);
	&alarm("sending etrn to $server");
	while (@hosts) {
		$server = shift(@hosts);
		&ps("etrn $server");
		while (<S>) {
			print if $verbose;
			last if /^\d+ /;
		}
		sleep(1);
	}
} else {
	print "\nETRN not supported\n\n"
}

&alarm("sending 'quit' to $server");
&ps("quit");
while (<S>) {
	print if $verbose;
	last if /^\d+ /;
}
close(S);
alarm(0);

select(STDOUT);
exit(0);

# print to the server (also to stdout, if -v)
sub ps
{
	my ($p) = @_;
	print ">>> $p\n" if $verbose;
	print S "$p\n";
}

sub alarm
{
	($alarm_action) = @_;
	alarm(10);
	$SIG{ALRM} = 'handle_alarm';
}

sub handle_alarm
{
	&giveup($alarm_action);
}

sub giveup
{
	my $reason = @_;
	(my $pk, my $file, my $line);
	($pk, $file, $line) = caller;

	print "Timed out during $reason\n" if $verbose;
	exit(1);
}

# read the rest of the current smtp daemon's response (and toss it away)
sub read_response
{
	(my $done, $verbose) = @_;
	(my @resp);
	print my $s if $verbose;
	while (($done eq "-") && ($s = <S>) && ($s =~ /^\d+([- ])/)) {
		print $s if $verbose;
		$done = $1;
		push(@resp, $s);
	}
	return @resp;
}
